### ISSUE TITLE
High CPU usage by highlighting code

### ISSUE THREAD

**Issue by 0xd4d** [rust-analyzer/rust-analyzer#3055]
Version: 3e1d97790be166f8735607c552a94a28ab9b09b8

Repro:

- open `main.rs` and `ee.rs` in VSCode
- switch to `main.rs` tab (thus hiding `ee.rs` tab)
- in `main.rs`'s `main()` func, type (don't copy and paste): `println!("hello world!");`
- monitor the CPU usage, one core is fully used for a long time

Files:

- lib.rs
```rust
pub mod ee;
```

- main.rs
```rust
fn main() {
    println!("Hello, world!");
    // 
}
```

- ee.rs
```rust
pub enum EE {
    E0,
    E1,
... add 3996 more lines here...
    E3998,
    E3999,
}
```

There appears to be different issues here:

1. If you just paste the `prinln!` example, it uses less CPU, so it seems to me that there's a lot of async code in the background, but they're not canceled once the user types another char.
2. `ee.rs` file is parsed (for the colorize feature?) even though it's not visible in the UI (it's opened in a tab but not visible). This seems to be a regression from RLS which does not seem to parse any files not visible in the UI. Is it possible to parse files (for colorize feature) only when they become visible in the UI?
3. Parsing all code to get color info for the editor should be delayed a little bit so it doesn't get triggered immediately. Maybe delay it by 200-500ms?

This shows that it's the highlighting code:

```
Line #, Process, Stack, Count, Weight (in view) (ms), TimeStamp (s), % Weight
5, ,   |- ra_lsp_server.exe!std::sys::windows::thread::{{impl}}::new::thread_start, 12241, 12,189.245000, , 18.20
6, ,   |    ra_lsp_server.exe!alloc::boxed::{{impl}}::call_once<(),FnOnce<()>>, 12241, 12,189.245000, , 18.20
7, ,   |    |- ra_lsp_server.exe!_ZN3std6thread7Builder5spawn17h3189275a84cf6bb7E, 12194, 12,142.284500, , 18.13
8, ,   |    |    ra_lsp_server.exe!panic_unwind::__rust_maybe_catch_panic, 12194, 12,142.284500, , 18.13
9, ,   |    |    ra_lsp_server.exe!_ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17hbdbe5d140016be3bE.llvm.12398782404485714031, 12194, 12,142.284500, , 18.13
10, ,   |    |    |- ra_lsp_server.exe!_ZN39_$LT$F$u20$as$u20$threadpool..FnBox$GT$8call_box17h0d31296badb3a9daE, 12171, 12,119.164200, , 18.09
11, ,   |    |    |    |- ra_lsp_server.exe!_ZN13ra_lsp_server9main_loop8handlers19publish_decorations17h966c7611245c95baE, 12110, 12,058.611900, , 18.00
12, ,   |    |    |    |    ra_lsp_server.exe!_ZN13ra_lsp_server9main_loop8handlers19publish_decorations17h966c7611245c95baE, 12110, 12,058.611900, , 18.00
13, ,   |    |    |    |    |- ra_lsp_server.exe!_ZN6ra_ide8Analysis9highlight17h0b8b4e78bb807426E, 12108, 12,056.642100, , 18.00
14, ,   |    |    |    |    |    ra_lsp_server.exe!_ZN5ra_db13CheckCanceled14catch_canceled17h1f22bb040a274324E, 12108, 12,056.642100, , 18.00
15, ,   |    |    |    |    |    ra_lsp_server.exe!panic_unwind::__rust_maybe_catch_panic, 12108, 12,056.642100, , 18.00
16, ,   |    |    |    |    |    ra_lsp_server.exe!_ZN3std9panicking3try7do_call17h403c34d712fbdd6fE.llvm.4113161438752224857, 12108, 12,056.642100, , 18.00
17, ,   |    |    |    |    |    ra_lsp_server.exe!_ZN6ra_ide19syntax_highlighting9highlight17hb6c1dc185997cd35E, 12108, 12,056.642100, , 18.00
18, ,   |    |    |    |    |    |- ra_lsp_server.exe!_ZN6ra_ide19syntax_highlighting9highlight17hb6c1dc185997cd35E, 12081, 12,029.793800, , 17.96
19, ,   |    |    |    |    |    |    |- ra_lsp_server.exe!_ZN9ra_ide_db4defs13classify_name17h08cec9caa8f9ae32E, 12069, 12,017.898100, , 17.94
20, ,   |    |    |    |    |    |    |    |- ra_lsp_server.exe!_ZN9ra_ide_db4defs15from_module_def17h986617082e528fb1E, 12018, 11,967.312300, , 17.86
21, ,   |    |    |    |    |    |    |    |    |- ra_lsp_server.exe!_ZN10ra_hir_def3adt80_$LT$impl$u20$ra_hir_def..src..HasChildSource$u20$for$u20$ra_hir_def..EnumId$GT$12child_source17h425f15acb4209916E, 8308, 8,273.077100, , 12.35
22, ,   |    |    |    |    |    |    |    |    |- ra_lsp_server.exe!_ZN13ra_hir_expand15InFile$LT$T$GT$3map17h2b6d5ae0023ea992E, 3701, 3,685.320900, , 5.50
```

**Comment by 0xd4d** [comment]
This is also a problem even if you don't have a big enum. Eg. if you have a lot of tabs open in VSCode, the CPU usage is much higher than if you only have one tab open.

**Comment by matklad** [comment]
The general problem here is that the LSP mixes together the "set of visible files" and "set of modified whose state on disk is outdated". The Dart protocol did this better, imo, but oh well. 

So, for highlighting we currently process all opened files. 

Hopefully this'll change once we swtich from our bespoke highlighting code to the current proposed API.

Regardless of this, it would be a good idea to investigate specific cases of slowness. rust-analyzer is not particularly optimized for CPU yet, but it makes sense to pick low hanging fruit!

**Comment by lnicola** [comment]
This seems reasonably fast now. I filed #5397 because completion seems annoyingly slow.
